lib/handlers/authorize-handler.js

```diff
@@ -0,0 +1,332 @@
+          var redirectUri = this.buildSuccessRedirectUri(uri, responseType);
+
+          this.updateResponse(response, redirectUri, state);
+
+          return code;
+        })
+        .catch(function(e) {
+          if (!(e instanceof OAuthError)) {
+            e = new ServerError(e);
+          }
+
+          var redirectUri = this.buildErrorRedirectUri(uri, e);
+
+          this.updateResponse(response, redirectUri, state);
+
+          throw e;
+        });
+    });
+};
+
+/**
+ * Generate authorization code.
+ */
+
+AuthorizeHandler.prototype.generateAuthorizationCode = function() {
+  if (this.model.generateAuthorizationCode) {
+    return Promise.try(this.model.generateAuthorizationCode);
+  }
+
+  return tokenUtil.generateRandomToken();
+};
+
+/**
+ * Get authorization code lifetime.
+ */
+
+AuthorizeHandler.prototype.getAuthorizationCodeLifetime = function() {
+  var expires = new Date();
+
+  expires.setSeconds(expires.getSeconds() + this.authorizationCodeLifetime);
+
+  return expires;
+};
+
+/**
+ * Get the client from the model.
+ */
+
+AuthorizeHandler.prototype.getClient = function(request) {
+  var clientId = request.body.client_id || request.query.client_id;
+
+  if (!clientId) {
+    throw new InvalidRequestError('Missing parameter: `client_id`');
+  }
+
+  if (!is.vschar(clientId)) {
+    throw new InvalidRequestError('Invalid parameter: `client_id`');
+  }
+
+  var redirectUri = request.body.redirect_uri || request.query.redirect_uri;
+
+  if (redirectUri && !is.uri(redirectUri)) {
+    throw new InvalidRequestError('Invalid request: `redirect_uri` is not a valid URI');
+  }
+
+  return Promise.try(this.model.getClient, clientId)
+    .then(function(client) {
+      if (!client) {
+        throw new InvalidClientError('Invalid client: client credentials are invalid');
+      }
+
+      if (!client.grants) {
+        throw new InvalidClientError('Invalid client: missing client `grants`');
+      }
+
+      if (!_.contains(client.grants, 'authorization_code')) {
+        throw new UnauthorizedClientError('Unauthorized client: `grant_type` is invalid');
+      }
+
+      if (!client.redirectUris || 0 === client.redirectUris.length) {
+        throw new InvalidClientError('Invalid client: missing client `redirectUri`');
+      }
+
+      if (redirectUri && !_.contains(client.redirectUris, redirectUri)) {
+        throw new InvalidClientError('Invalid client: `redirect_uri` does not match client value');
+      }
+
+      return client;
+    });
+};
+
+/**
+ * Get scope from the request.
+ */
+
+AuthorizeHandler.prototype.getScope = function(request) {
+  var scope = request.body.scope || request.query.scope;
+
+  if (!is.nqschar(scope)) {
+    throw new InvalidScopeError('Invalid parameter: `scope`');
+  }
+
+  return scope;
+};
+
+/**
+ * Get state from the request.
+ */
+
+AuthorizeHandler.prototype.getState = function(request) {
+  var state = request.body.state || request.query.state;
+
+  if (!state) {
```

Comment:
I understand but let me be honest here: this entire PR is a major breaking change. There's no way our users will simply upgrade their `node-oauth2-server` dependency to `3.0.0` and walk away, so why not take this unique chance to educate our users?

I don't mean to force my views on this, but [it's](http://www.thread-safe.com/2014/05/the-correct-use-of-state-parameter-in.html) [not](http://www.twobotechnologies.com/blog/2014/02/importance-of-state-in-oauth2.html) [just](http://homakov.blogspot.pt/2012/07/saferweb-most-common-oauth2.html) [me](http://spring.io/blog/2011/11/30/cross-site-request-forgery-and-oauth2).

If I'm alone in this argument I'll make it overridable but I'd like to state for the record that I think the `state` parameter should absolutely **not** be optional.